src/layout/Popup.js

```diff
@@ -0,0 +1,138 @@
+    Popup.prototype.publish("buttonName", null, "string", "Button name property",null,{});
+    Popup.prototype.publish("buttonLabel", "", "string", "Button label text",null,{});
+    Popup.prototype.publish("widget", null, "widget", "Widget",null,{tags:['Private']});
+
+
+    Popup.prototype.testData = function () {
+        this.showTestButton(true);
+        this.position("absolute");
+        this.top(30);
+        
+        var context = this;
+        this._testButton = new Input()
+            .type("button")
+            .value("Open Popup")
+            .on("click", function () {
+                context.updateState(!(context.popupState()));
+            }, true)
+        ; 
+
+        this.widget(new Surface().widget(new Column().testData()));
+
+        return this;
+    };
+
+    Popup.prototype.widgetSize = function (widgetDiv) {
+        var width = this.clientWidth() - this.calcFrameWidth(widgetDiv);
+        var height = this.clientHeight() - this.calcFrameHeight(widgetDiv);
+ 
+        return { width: width, height: height };
+    };
+    
+    Popup.prototype.updateState = function (visible) {
+            this.popupState(visible).render();
+    };
+    
+    Popup.prototype.enter = function (domNode, element) {
+        HTMLWidget.prototype.enter.apply(this, arguments);
+
+        if (this._testButton) {
+            this._testButton.target(domNode.parentNode).render();
+        }
+    };
+
+    Popup.prototype.update = function (domNode, element) {
+        HTMLWidget.prototype.update.apply(this, arguments);
+        var context = this;        
+
+        element
+            .style("position",this.position())
+            .style("left",this.left() + "px")
+            .style("right",this.right() + "px")
+            .style("top",this.top() + "px")
+            .style("bottom",this.bottom() + "px")
+        ;
+        
+        if (this.popupState()) {
+            element.style("display", "block");
+        } else {
+            element.style("display", "none");
+        }
+        
+        var widgets = element.selectAll("#" + this._id + " > .popupWidget").data(this.widget() ? [this.widget()] : [], function (d) { return d._id; });
+        widgets.enter().append("div")
+            .attr("class", "popupWidget")
+            .each(function (d) {
+                d.target(this);
+            })
+        ;
+        widgets
+            .each(function (d) {
+                var widgetSize = context.widgetSize(d3.select(this));
+                d.resize({ width: widgetSize.width, height: widgetSize.height });
+            })
+        ;
+        widgets.exit().each(function (d) {
+            d.target(null);
+        }).remove();
+        
+    };
+
+    Popup.prototype.exit = function (domNode, element) {
+        if (this.widget()) {
+            this.widget(null);
+            this.render();
+        }
+        HTMLWidget.prototype.exit.apply(this, arguments);
+    };
+
+    Popup.prototype.render = function (callback) {
```

Comment:
is there a diff way to handle the callback, as i have a PR for a new kind of callback method and this would interfere with it potentially https://github.com/hpcc-systems/Visualization/pull/602  ... its still very much debatable if it will get used though